packages/driver/src/cypress/ensure.ts

```diff
@@ -0,0 +1,314 @@
+import _ from 'lodash'
+import $dom from '../dom'
+import $utils from './utils'
+import $errUtils from './error_utils'
+import type { $Cy } from './cy'
+import { isRunnerAbleToCommunicateWithAut } from '../util/commandAUTCommunication'
+
+// TODO: in 4.0 we should accept a new validation type called 'elements'
+// which accepts an array of elements (and they all have to be elements!!)
+// this would fix the TODO below, and also ensure that commands understand
+// they may need to work with both element arrays, or specific items
+// such as a single element, a single document, or single window
+
+// TODO: we should probably normalize all subjects
+// into an array and loop through each and verify
+// each element in the array is valid. as it stands
+// we only validate the first
+const validateType = (subject, type, name: string, cy: $Cy) => {
+  switch (type) {
+    case 'element':
+      // if this is an element then ensure its currently attached
+      // to its document context
+      if ($dom.isElement(subject)) {
+        isAttached(subject, name, cy)
+      }
+
+      // always ensure this is an element
+      return isElement(subject, name, cy)
+
+    case 'document':
+      return isDocument(subject, name, cy)
+
+    case 'window':
+      return isWindow(subject, name, cy)
+
+    default:
+      return
+  }
+}
+
+const isType = (subject, type, name: string, cy: $Cy) => {
+  let types: (string | boolean)[] = [].concat(type)
+
+  // if we have an optional subject and nothing's
+  // here then just return cuz we good to go
+  if (types.includes('optional') && _.isUndefined(subject)) {
+    return
+  }
+
+  // okay we either have a subject and either way
+  // slice out optional so we can verify against
+  // the various types
+  types = _.without(types, 'optional')
+
+  // if we have no types then bail
+  if (types.length === 0) {
+    return
+  }
+
+  let err
+  const errors: Error[] = []
+
+  for (type of types) {
+    try {
+      validateType(subject, type, name, cy)
+    } catch (error) {
+      err = error
+      errors.push(err)
+    }
+  }
+
+  function wordJoin (array) {
+    const copy = [...array]
+    const last = copy.pop()
+
+    return `${copy.join(', ')} or ${last}`
+  }
+
+  // every validation failed and we had more than one validation
+  if (errors.length === types.length) {
+    err = errors[0]
+
+    if (types.length > 1) {
+      // append a nice error message telling the user this
+      const msg = err.message.replace(/(failed because it requires .*)\./, `${wordJoin(['$1', ...types.slice(1)]) }.`)
+      const errProps = $errUtils.modifyErrMsg(err, '', () => msg)
+
+      $errUtils.mergeErrProps(err, errProps)
+    }
+
+    throw err
+  }
+}
+
+const isChildCommand = (command, args) => {
+  if (cy.subjectChain(command.get('chainerId')) === undefined) {
```

Comment:
The other helpers that leverage `cy` have them passes as an argument. Should we here?

Reply:
Yeah, definitely. Updated.